src/firmware.cpp

```diff
@@ -60,118 +42,104 @@ static String pop_option(String& argument) {
   }
 }
 
-static CommandError run_help(int commandId, String argument);
+static void serialWrite(int requestID, char lineType, const String& str) {
+    if (requestID != 0) {
+        Serial.write('@');
+        Serial.print(requestID, DEC);
+        Serial.write(' ');
+    }
+
+    Serial.write(lineType);
+    Serial.write(' ');
 
-static CommandError led(int commandId, String argument) {
-  if (argument == "on") {
-    digitalWrite(LED_BUILTIN, HIGH);
-  } else if (argument == "off") {
-    digitalWrite(LED_BUILTIN, LOW);
-  } else {
-    return COMMAND_ERROR("unknown argument");
-  }
-  return OK;
+    Serial.println(str);
 }
 
-static CommandError servo(int commandId, String argument) {
-  String servoArg = pop_option(argument);
-  String widthArg = pop_option(argument);
+// The actual commands
 
-  if (argument.length() || !servoArg.length() || !widthArg.length()) {
-    return COMMAND_ERROR("servo takes exactly two arguments");
-  }
+static void readAnaloguePin(int requestID, const String& name, int pin) {
+  int reading = analogRead(pin);
+  serialWrite(requestID, '>', name + " " + String(reading));
+}
 
-  auto width = widthArg.toInt();
-  auto servo = servoArg.toInt();
-  if (servo < 0 || servo > 15) {
-    return COMMAND_ERROR("servo index out of range");
-  }
-  if (width != 0 && (width < 150 || width > 550)) {
-    return COMMAND_ERROR("width must be 0 or between 150 and 550");
-  }
-  SERVOS.setPWM(servo, 0, width);
+static CommandResponse analogueRead(int requestID, String argument) {
+  readAnaloguePin(requestID, "a0", A0);
+  readAnaloguePin(requestID, "a1", A1);
+  readAnaloguePin(requestID, "a2", A2);
+  readAnaloguePin(requestID, "a3", A3);
   return OK;
 }
 
-static CommandError write_pin(int commandId, String argument) {
-  String pinIDArg = pop_option(argument);
-  String pinStateArg = pop_option(argument);
-
-  if (argument.length() || !pinIDArg.length() || !pinStateArg.length()) {
-    return COMMAND_ERROR("need exactly two arguments: <pin> <high/low/hi-z/pullup>");
-  }
-
-  int pin = pinIDArg.toInt();
-
-  if (pin < 2 || pin > 12) {
-    return COMMAND_ERROR("pin must be between 2 and 12");
-  }
-
-  if (pinStateArg == "high") {
-    pinMode(pin, OUTPUT);
-    digitalWrite(pin, HIGH);
-  } else if (pinStateArg == "low") {
-    pinMode(pin, OUTPUT);
-    digitalWrite(pin, LOW);
-  } else if (pinStateArg == "hi-z") {
-    pinMode(pin, INPUT);
-  } else if (pinStateArg == "pullup") {
-    pinMode(pin, INPUT_PULLUP);
+static CommandResponse led(int requestID, String argument) {
+  if (argument == "H") {
```

Comment:
This might work better as a `switch`?

Reply:
I see no reason why it wouldn't, i'd expect the `switch` to use `==` for value comparison anyway

Reply:
This can be changed to a switch by getting the first `char` of the argument. I personally dislike using `String` in `switch` anyway.

Reply:
Fixed in 6c67c51